quickwit/quickwit-indexing/src/source/queue_sources/processor.rs

```diff
@@ -0,0 +1,464 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt;
+use std::sync::Arc;
+use std::time::Duration;
+
+use quickwit_actors::{ActorExitStatus, Mailbox};
+use quickwit_config::QueueParams;
+use quickwit_metastore::checkpoint::SourceCheckpoint;
+use quickwit_proto::indexing::IndexingPipelineId;
+use quickwit_proto::metastore::SourceType;
+use quickwit_storage::StorageResolver;
+use serde_json::{json, Value as JsonValue};
+use tracing::debug;
+use ulid::Ulid;
+
+use super::local_state::QueueLocalState;
+use super::message::CheckpointedMessage;
+use super::shared_state::{QueueSharedState, QueueSharedStateImpl};
+use super::visibility::{spawn_visibility_task, VisibilityTaskHandle};
+use super::{acknowledge, Queue};
+use crate::actors::DocProcessor;
+use crate::models::{NewPublishLock, NewPublishToken, PublishLock};
+use crate::source::{SourceContext, SourceRuntime};
+
+#[derive(Default)]
+pub struct QueueProcessorObservableState {
+    /// Number of bytes processed by the source.
+    num_bytes_processed: u64,
+    /// Number of messages processed by the source.
+    num_messages_processed: u64,
+    // Number of invalid messages, i.e., that were empty or could not be parsed.
+    num_invalid_messages: u64,
+    /// Number of time we looped without getting a single message
+    num_consecutive_empty_batches: u64,
+}
+
+pub struct QueueProcessor {
```

Comment:
Processor is not a great name.... we already use it do describe the thing that transforms json into a document.

Reply:
It is now a `QueueCoordinator`